### PR TITLE
Improve readability for server_test

### PR DESCRIPTION
**Pull request by KianYang-Lee**
Use ready chan to signal that server are ready to accept to accept connection in tests. Included tests for WithReadinessChan and modify test coverage for ErrServerIsRunning.

### CODE REVIEW

File: server/server_test.go [hunk 1]
```diff
@@ -70,55 +71,82 @@ func TestServer_WithBaseContext(t *testing.T) {
...
+					close(done)
+				case ErrServerAlreadyRunning:
+					return
+				default:
+					t.Errorf("Got unexpected error: %v", err)
+				}
+			}()
+
+			// Wait for server to be ready
+			<-ready
```
Comment (ksysoev) [inline server/server_test.go:123]: please cover this waiting with select that has timeout as alternative option, to not dead lock if ready channel is broken

```
select {
 			case <-ready:
 			case <-time.After(1 * time.Second):
 				t.Error("Expected server to start")
 			}
```

File: server/server_test.go [hunk 2]
```diff
@@ -70,55 +71,82 @@ func TestServer_WithBaseContext(t *testing.T) {
...
-		go func() {
-			err := server.Run()
-			switch err {
-			case nil:
-				close(done)
-			case ErrServerAlreadyRunning:
-				done <- struct{}{}
-			default:
-				t.Errorf("Got unexpected error: %v", err)
+	noOfReruns := []int{0, 1, 2, 5}
```
Comment (ksysoev) [inline server/server_test.go:92]: not sure it gives additional value

File: server/server_test.go [hunk 3]
```diff
@@ -189,28 +207,19 @@ func TestServer_Close_NoContext(t *testing.T) {
...
+			done <- struct{}{}
+		default:
+			t.Errorf("Got unexpected error: %v", err)
 		}
-	case <-time.After(1 * time.Second):
-		t.Error("Expected server to start")
-	}
+	}()
+
+	<-ready
```
Comment (ksysoev) [inline server/server_test.go:222]: same here, please add timeout handling

File: server/server_test.go [hunk 4]
```diff
@@ -137,29 +165,18 @@ func TestServer_Close(t *testing.T) {
...
-	select {
-	case _, ok := <-done:
-		if !ok {
-			t.Error("Expected server to start")
+	go func() {
+		err := server.Run()
+		switch err {
+		case nil:
+			close(done)
+		case ErrServerAlreadyRunning:
```
Comment (ksysoev) [inline server/server_test.go:173]: I think this case is not needed anymore
```suggestion
```

File: server/server_test.go [hunk 5]
```diff
@@ -189,28 +207,19 @@ func TestServer_Close_NoContext(t *testing.T) {
...
+			done <- struct{}{}
```
Comment (ksysoev) [inline server/server_test.go:216]: looks to be dead code
```suggestion
```

File: server/server_test.go [hunk 6]
```diff
@@ -137,29 +165,18 @@ func TestServer_Close(t *testing.T) {
...
+		case ErrServerAlreadyRunning:
+		default:
+			t.Errorf("Got unexpected error: %v", err)
 		}
-	case <-time.After(1 * time.Second):
-		t.Error("Expected server to start")
-	}
+	}()
 
+	<-ready
```
Comment (ksysoev) [inline server/server_test.go:179]: + timeout handling

**Review by ksysoev — CHANGES_REQUESTED** [review]
@KianYang-Lee Thanks, tests are looking better with readiness channel, left few comments.

Also could you check your git config on your machine(email and name), looks GitHub not able to map your commits with your GitHub account and not running CI for your commits.

Also please could you check that code passing linter rules  https://github.com/ksysoev/wasabi/blob/main/CONTRIBUTING.md#get-the-code-linter-running

File: server/server_test.go (hunk 2 above)
Comment (KianYang-Lee) [reply server/server_test.go:92]: calling Run initially should not throw error, but repeatedly should throws error, this check test for that scenario. 5 might be unnecessary though so I removed it

**Review by ksysoev — APPROVED** [review]
LGTM

**Pull request merged**